toolchain/check/context.h

```diff
@@ -383,8 +383,9 @@ class Context {
   auto GetAssociatedEntityType(SemIR::TypeId interface_type_id,
                                SemIR::TypeId entity_type_id) -> SemIR::TypeId;
 
-  // Gets a builtin type. The returned type will be complete.
-  auto GetBuiltinType(SemIR::BuiltinInstKind kind) -> SemIR::TypeId;
+  // Gets a singleton type. The returned type will be complete. Requires that
+  // `singleton_id` is already validated to be a singleton.
+  auto GetSingletonType(SemIR::InstId singleton_id) -> SemIR::TypeId;
```

Comment:
Have you considered passing the singleton type as a template argument instead of passing the instruction ID? Eg:
```c++
  template <typename SingletonT> requires IsSingletonInstId(SingletonT::SingletonInstId)
  auto GetSingletonType() -> SemIR::TypeId {
    return GetSingletonType(SingletonT::SingletonInstId);
  }
```
That'd allow us to do the error checking at compile time and make the calls a bit less repetitive:
```diff
-        loc_id, {.type_id = context.GetSingletonType(
-                     SemIR::IntLiteralType::SingletonInstId),
+        loc_id, {.type_id = context.GetSingletonType<SemIR::IntLiteralType>(),
```

Reply:
We have runtime calls such as:

```
      return local_context().GetSingletonType(import_type_inst_id);
```

Does this still seem worth it to you if it means we need two separate implementations?

Reply:
Yes, I wasn't really a big fan of the wrapper function. GetSingletonType(InstId) can't be removed because it's used at runtime.